### DISCUSSION TITLE
How to avoid an exception ... or ... do a word exist?

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#126] [General]
Hi Travis,

Currently I am writing words into a file with zeptoedit, save the file, "inluded" the file, test the words, correct bug while editing the file in zed and so on.

To not to explode the dictionary with the same word defintion over and over again :)  and to keep things lean I start the file
with the sequence

markername
marker markername

which erases the previous version and created a new marker for the next version of words.

This words as long as I set a marker "markername" when I start programming. Otherwise I get an unknown-word exception.

Is there a way to check, whether a word exists without triggering an error/exception?

Another way would be to surround the first line like so

try
 markername
suppress 

Unfortunately I can't find an example for using 'suppress'. I searched "Getting starting" (chapter "exceptions") and grepped through the source of zeptoforth.

What is the correct syntax for try/suppress?

Cheers!
Tuxic!

PS: Thanks a lot for release 1.7.3 ! Will be flashed in the moments! :)

**Comment by tabemann** [comment]
On Fri, Aug 2, 2024 at 9:25 PM mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> Currently I am writing words into a file with zeptoedit, save the file,
> "inluded" the file, test the words, correct bug while editing the file in
> zed and so on.
>
> To not to explode the dictionary with the same word defintion over and
> over again :) and to keep things lean I start the file
> with the sequence
>
> markername
> marker markername
>
> which erases the previous version and created a new marker for the next
> version of words.
>
> This words as long as I set a marker "markername" when I start
> programming. Otherwise I get an unknown-word exception.
>
> Is there a way to check, whether a word exists without triggering an
> error/exception?
>

What you are looking for is

defined? markername [if] markername [then]
marker markername

Note that this will only work when compiling to RAM, because if you are
compiling to flash executing MARKERNAME will reboot the MCU, disconnecting
your USB connection.


> Another way would be to surround the first line like so
>
> try
> markername
> suppress
>
> Unfortunately I can't find an example for using 'suppress'. I searched
> "Getting starting" (chapter "exceptions") and grepped through the source of
> zeptoforth.
>
> What is the correct syntax for try/suppress?
>
> Cheers!
> Tuxic!
>

TRY won't help you here, because the error message generated when
attempting to compile or call a word not in the current search order is
generated at compilation or interpretation time, and TRY expects an xt
which it calls and which may or may not raise an exception at runtime. For
instance if you did

:noname markername ; try drop

you would get an exception while compiling the anonymous word, before TRY
could be called.

As for SUPPRESS, it is an immediate word that can be used both in
interpretation and compilation mode, and has the signature ( exc|0 "name"
-- exc|0 ) where "name" is the name (without quotes) of the exception you
want to suppress, and exc is the exception that may or may not be that
exception; if it matches then 0 is returned, otherwise it is a
pass-through. However, it is not what you want here.

PS: Thanks a lot for release 1.7.3 ! Will be flashed in the moments! :)
>

You're welcome!

Travis

**Comment by mcctuxic** [comment]
Hi Travis,

thank you! That helps me a lot! Still I am not thinking forthic enough :)

Cheers!
Tuxic

**Comment by tabemann** [comment]
On Fri, Aug 2, 2024 at 10:35 PM mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> thank you! That helps me a lot! Still I am not thinking forthic enough :)
>
> Cheers!
> Tuxic
>

Well, DEFINED? isn't a standard word, but one I created for my own purposes
because I found I needed to have conditional compilation based upon whether
a given word was defined.

Note that if you want to use DEFINED? inside a word definition, e.g. with
[IF] ... [THEN] or [IF] ... [ELSE] ... [THEN], you need to do something
like the following:

: foo [ defined? bar ] [if] bar [then] ;

because DEFINED? is not a special compilation word, i.e. it does not
specially handle its argument token at compile-time, and [IF] takes its
argument off the data stack at interpretation/compilation time. As a
result, you need to evaluate DEFINED? at interpretation time, hence the [
... ].

Travis

**Comment by mcctuxic** [comment]
Hi Travis,

defined? is super handy  ...and I search the docs for the word "exists?", hahahaha!
I will rethink the []-thing ... probably resulting in more questions...

Currently I am (trying to) write some code to drive a nixie clock electronic via SPI. Each tube has
a little driver board (I definitely didn't want to handle a high voltage in the neighbourhood of a microcontroller myself ;) )
Beside the nixie tube itsself the driver board has a RGB led.
But I got stuck while trying to define a human readable/usable word:
I need to set the brightness value (0...127) and to enable/disable the according led (set/clear 128)
of either byte 13,14 or 15 of the 16 byte long spi buffer. I have already defined a word, which
does that. But 
13 55 lit-led 
isn't very readable I think. I would like to do something like
55 lit-led Red
...but I cannot get my head around the "how can I read a character/string, which is _after_ the defined word" thing.
I litterally have no words.... :)

Cheers!
Tuxic

**Comment by tabemann** [comment]
Why not 55 red lit-led ? In this case you probably do not want to read
after LIT-LED . Sure there is TOKEN but I do not recommend it here. (You
would have to make LIT-LED immediate and do compiler magic with POSTPONE
and LIT, to get it to work and it would be less flexible because you would
not be able to pass around a color then.)

Travis

**Comment by mcctuxic** [comment]
Oh! :) OF COURSE! Yes, that makes much more sense! It is easy to define
constants for that. Thank you!

**Comment by tabemann** [comment]
On Sat, Aug 3, 2024 at 12:52 AM mcctuxic ***@***.***> wrote:

> Oh! :) OF COURSE! Yes, that makes much more sense! It is easy to define
> constants for that. Thank you!
>
>
> On 08/02 10:46, tabemann wrote:
> > Why not 55 red lit-led ? In this case you probably do not want to read
> > after LIT-LED . Sure there is TOKEN but I do not recommend it here. (You
> > would have to make LIT-LED immediate and do compiler magic with POSTPONE
> > and LIT, to get it to work and it would be less flexible because you
> would
> > not be able to pass around a color then.)


Generally, a solution where you don't have to use TOKEN or TOKEN-WORD
(which is essentially TOKEN combined with FIND * with error-checking) is
better than one in which you do, if one is available. There are a few
places where I break this rule, such as in AVERTS or TRIGGERS (where in
that case it is as syntactic sugar to avoid the user having to specify [']
repeatedly), but in general I try to follow it.

* Note - this FIND is not like the ANS Forth FIND in that it looks up a
word header rather than an xt.

Travis

**Comment by mcctuxic** [comment]
(I am asking to check whether I have understood the underlyong principle...)
Avoiding this <value> <word> <token> construct is possible only of <token> "is"
already someting - a defined constant for example. If one creates "something
new with a name of <token>" there is no way around this, because otherwise
the interpreter will fall over an undefined token/word, isn't it.
"100 variable onehundred" would not work when used as
100 onehundred variable, due to the unknown'ness' of 'onehundred'.
(?)

**Comment by tabemann** [comment]
On Sat, Aug 3, 2024, 09:58 mcctuxic ***@***.***> wrote:

> (I am asking to check whether I have understood the underlyong
> principle...)
> Avoiding this construct is possible only of "is"
> already someting - a defined constant for example. If one creates
> "something
> new with a name of " there is no way around this, because otherwise
> the interpreter will fall over an undefined token/word, isn't it.
> "100 variable onehundred" would not work when used as
> 100 onehundred variable, due to the unknown'ness' of 'onehundred'.
> (?)
>
Basically yes. If you have a value that is already known, including the xt
of a word, you normally pass it in on the data stack. If you have a name
which you want to define/declare, you have a name that you want to look up,
or you have literal data you want to do something special with you then
should place it after the word in question and use something like TOKEN to
get.

Travis

>
